crates/pera-core/src/dwallet_mpc/mpc_session.rs

```diff
@@ -0,0 +1,212 @@
+use std::collections::HashMap;
+use std::sync::{Arc, Weak};
+
+use group::PartyID;
+use mpc::{AsynchronousRoundResult, WeightedThresholdAccessStructure};
+
+use dwallet_mpc_types::dwallet_mpc::{MPCMessage, MPCPublicInput, MPCSessionStatus};
+
+use pera_types::base_types::EpochId;
+use pera_types::dwallet_mpc_error::{DwalletMPCError, DwalletMPCResult};
+use pera_types::messages_consensus::ConsensusTransaction;
+use pera_types::messages_dwallet_mpc::SessionInfo;
+
+use crate::authority::authority_per_epoch_store::AuthorityPerEpochStore;
+use crate::dwallet_mpc::mpc_party::MPCParty;
+use crate::dwallet_mpc::{authority_name_to_party_id, DWalletMPCMessage};
+
+/// A dWallet MPC session.
+/// It keeps track of the session, the channel to send messages to the session,
+/// and the messages that are pending to be sent to the session.
+pub(super) struct DWalletMPCSession {
+    /// The status of the MPC session.
+    pub(super) status: MPCSessionStatus,
+    /// The messages that are pending to be executed while advancing the session
+    /// We need to accumulate a threshold of those before advancing the session.
+    pub(super) pending_messages: Vec<HashMap<PartyID, Vec<u8>>>,
+    epoch_store: Weak<AuthorityPerEpochStore>,
+    epoch_id: EpochId,
+    /// The total number of parties in the chain
+    /// We can calculate the threshold and parties IDs (indexes) from it.
+    /// To calculate the party's ID, all we need to know is the number of parties,
+    /// as the IDs are just the indexes of those parties.
+    /// If there are three parties, the IDs are [0, 1, 2].
+    pub(super) session_info: SessionInfo,
+    /// The MPC party being used to run the MPC cryptographic steps.
+    /// Party in here is not a Validator, but a cryptographic party.
+    party: MPCParty,
+    pub(super) public_input: MPCPublicInput,
+}
+
+/// Needed to be able to iterate over a vector of generic DWalletMPCSession with Rayon.
+unsafe impl Send for DWalletMPCSession {}
```

Comment:
Please no `unsafe`. `Send` and `Sync` are automatically derived for types that support them, so if you need this manual implementation it probably means it is actually unsafe. Please check all members of the struct and figure out why it wasn't derived automatically